core/Pos/Binary/Class/Primitive.hs

```diff
@@ -168,3 +174,34 @@ fromBinaryM = either (fail . toString) return . fromBinary
 biSize :: Bi a => a -> Byte
 biSize = fromIntegral . BS.length . serialize'
 {-# INLINE biSize #-}
+
+----------------------------------------------------------------------------
+-- CBORDataItem
+-- https://tools.ietf.org/html/rfc7049#section-2.4.4.1
+----------------------------------------------------------------------------
+
+-- | Encode the given `ByteString` and sorrounds it with the semantic tag 24.
+-- In CBOR diagnostic notation:
+-- >>> 24(h'DEADBEEF')
+encodeCborDataItem :: ByteString -> E.Encoding
+encodeCborDataItem bs = E.encodeTag 24 <> encode bs
+
+-- | Remove the the semantic tag 24 from the enclosed CBOR data item,
+-- failing if the tag cannot be found.
+decodeCborDataItemTag :: D.Decoder s ()
+decodeCborDataItemTag = do
+    t <- D.decodeTag
+    when (t /= 24) $ fail $
+        "decodeCborDataItem: expected a bytestring with \
+        \CBOR (marked by tag 24), found tag: " <> show t
+
+-- | Remove the the semantic tag 24 from the enclosed CBOR data item,
+-- decoding back the inner `ByteString` as a proper Haskell type.
+-- Consume its input in full.
+decodeCborDataItem :: Bi a => D.Decoder s a
+decodeCborDataItem = do
+    decodeCborDataItemTag
+    bs <- decode @ByteString
+    case decodeFull bs of
+        Left e  -> fail (toString e)
+        Right v -> pure v
```

Comment:
These utils look correct. But note that they are not symmetric. The encode takes a ByteString (ie does no extra encoding) but the decode side does an extra decode step, with the `Bi a` constraint.

It would probably be easier to follow below if the two sides were consistent with each other.

Perhaps that means you need to split it into two cases: a known case that uses a `Bi a` constraint on encode and decode, and an unknown case that uses `ByteString`.

```
encodeKnownCborDataItem :: Bi a => a -> E.Encoding
encodeKnownCborDataItem x = E.encodeTag 24 <> encode (serialise' x)

encodeUnknownCborDataItem :: ByteString -> E.Encoding
encodeUnknownCborDataItem bs = E.encodeTag 24 <> encode bs
```

And correspondingly for the decode side. The existing `decodeCborDataItem` would just be renamed to `decodeKnownCborDataItem` and you'd add the `decodeUnknownCborDataItem` case.

And once you have these I don't think you'll need to export `decodeCborDataItemTag`.